File: Day248/champagneTower.java
package Day248;


import java.util.ArrayList;
import java.util.List;

public class champagneTower {

    public static double tower(int poured, int query_row, int query_glass) {
        // List<List<Double>> list = new ArrayList<>();
        if (poured ==  0 ){
            return 0;
        }
        List<Double> list = new ArrayList<>();
        list.add( new Double(poured));
        while (query_row-- > 0) {
            List<Double> temp = new ArrayList<>();
            temp.add(Math.max((list.get(0)-1)/2, 0));
            for (int i=1; i<list.size(); i++) {
                temp.add(Math.max((list.get(i-1)-1)/2, 0) + Math.max((list.get(i)-1)/2, 0));
            }
            temp.add(temp.get(0));
            list = new ArrayList<>(temp);
        }
        return Math.min(1, list.get(query_glass));
    }

    public static void main(String[] args) {
        int poured = 2;
        int query_row = 1;
        int query_glass = 1;
        System.out.println(tower(poured, query_row, query_glass));
    }
}

File: Day248/champagneTower.go
package main

import (
	"fmt"
	"math"
)

func champagneTower(poured int, query_row int, query_glass int) float64 {
	tower := make([][]float64, query_row+1)
	for i := range tower {
		tower[i] = make([]float64, i+1)
	}

	tower[0][0] = float64(poured)

	for i := 0; i < query_row; i++ {
		for j := 0; j <= i; j++ {
			excess := (tower[i][j] - 1) / 2.0
			if excess > 0 {
				tower[i+1][j] += excess
				tower[i+1][j+1] += excess
			}
		}
	}

	return math.Min(1.0, tower[query_row][query_glass])
}

func main() {
	poured := 2
	query_row := 1
	query_glass := 1
	fmt.Println(champagneTower(poured, query_row, query_glass))
}
